### ISSUE TITLE
Worldserver hangs when running docker-compose up

### ISSUE THREAD

**Issue by LunarEngineer** [JustinChristensen/trinitycore-dockerfile#5]
System Details
Docker Version: Docker version 19.03.12, build 48a66213fe
Operation System: Ubuntu (uname -a: Linux gaia 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux)
Any other details relevant to the problem.

This is running the stock implementation directly from your github repository, with the commit from your Github repo readme.md.

When I run the vanilla docker-compose flow from your repo (i.e. clone and run the shell script to start things up) everything works great for the auth and db containers. The world container builds but console output stops at 
```
...
worldserver_1  | >> Loaded 16 faction change reputation pairs in 0 ms
worldserver_1  | Loading faction change title pairs...
```

It just hangs there.

I know that it didn't run to completion because I cannot log in to the server from a client, though the RA port functions fine and I can telnet in.

If I open a shell inside the container and run the worldserver it runs to completion (though I ran it with the other worldserver still hung, so it died when it tried to seize the ports at the end of the world.cpp initialization routine, but it made it to that point without issues.)

This is behavior that I saw when I was running my own version (though hung at a different point), with modifications, so I tested your vanilla version to see if I had introduced errors and I'm seeing the same behavior.

I'm frankly a little baffled by this and I'm not sure why it's happening, though I am inclined to lean towards something with the way that Docker pipes to standard out.

I will be testing redirecting the standard out to file to see if that fixes the problem.

**Comment by JustinChristensen** [comment]
That's not a hang. That's the last message it prints. It would be an improvement if the WorldServer printed a message announcing that startup was successful, but it doesn't.

It appears I overlooked a critical step. You'll need to configure your WoW client by following the instructions here:
https://trinitycore.atlassian.net/wiki/spaces/tc/pages/74006268/Client+Setup

If I'm feeling plucky, I may either update the README or add a script to automate this, but for now you'll just need to do it manually.

Let me know if this can be closed.

**Comment by LunarEngineer** [comment]
Unfortunately the client is configured correctly. I'm transitioning my server from a VM to Docker so I'm moderately comfortable with Trinity.

I am pretty sure I tested the client correctly because the auth server handles the traffic and I can see that the world exists, I just cannot log in.

I will test again when I'm off work today to double check.

**Comment by JustinChristensen** [comment]
To verify that the world server is up and running and the server is listening, use a tool like netcat or telnet to test a connection to port 8085 on the host. If the test succeeds it's likely an issue with the client's configuration.

**Comment by LunarEngineer** [comment]
TL;DR: I did something silly and I should have known better.

Slightly longer story: I run my server on the beast in the basement and was trying to connect from my laptop client. In the realmlist table the IP address for the worldserver is still 127.0.0.1. The IP routing was cheerfully trying to tell me that there was no world server on my laptop. Kind of a client issue, but not really?

The worldserver runs fine and when I pay attention to which IP address it should be connecting to everything is just peachy.

Bottom line is that this can be closed (and I will), though the issue itself is unresolved; the issue is unrelated to the Trinity server and does not affect anything other than my sensibilities.

Looking at the [world init routine](https://github.com/TrinityCore/TrinityCore/blob/c56d0a092fa5732d5131713e355ae850d672e090/src/server/game/World/World.cpp) the last string that should be printed to stdout is "World initialized in %u minutes %u seconds".

This particular string that the CLI output is frozen at is halfway through the init routine.

I'd still like to hunt this down because it bothers me. Any ideas on where I should start chasing it down? My suspicion is still the connection between Docker and STDOUT. Feel free to just ignore this.

I would also like to say *thank you* for being responsive. I realize this project is something you're not maintaining anymore and I also realize how precious time can be.

**Comment by JustinChristensen** [comment]
Likely output buffering. Now that you mention it, I did encounter this. You'll have to dig to find a solution.

**Comment by JustinChristensen** [comment]
I did look into this a bit this morning, and the issue is that stdout is fully buffered when connected to a pipe, which is the default for `docker-compose up`. If you run with `docker-compose run` or add `tty: true` to the worldserver and authserver services in docker-compose.yaml, it's connected to a TTY and line-buffered like normal.

So a quick hacky fix would be to force a pseudo-TTY for each service in docker-compose.yaml (with `tty: true`). 

A program can control buffering by setting the buffering mode for a given file descriptor, but TrinityCore isn't doing anything special as far as that goes. I suspect MariaDB is explicitly detecting details about the device and setting the buffering mode (so that the output is the same regardless of the connected device).

HTH.

**Comment by LunarEngineer** [comment]
That does help a lot, it's very appreciated!
